torch/distributed/checkpoint/state_dict.py

```diff
@@ -445,17 +459,35 @@ def _load_model_state_dict(
     state_dict: Dict[str, ValueType],
     info: _StateDictInfo,
 ) -> _IncompatibleKeys:
-    if not info.handle_model or not state_dict:
+    if not info.handle_model or (not state_dict and not info.broadcast_from_rank0):
         return _IncompatibleKeys({}, {})
 
-    for key, _ in _iterate_valid_model_state(model):
+    local_state_dict = {}
+    for key, value in _iterate_valid_model_state(model):
         fqns = _get_fqns(model, key)
         fqns_with_prefix = _get_fqns(
             model, key, skip_ddp_prefix=False, skip_compiler_prefix=False
         )
+
         for fqn, fqn_with_prefix in zip(fqns, fqns_with_prefix):
-            if fqn != fqn_with_prefix:
+            if (
+                not info.broadcast_from_rank0 or dist.get_rank() == 0
+            ) and fqn != fqn_with_prefix:
                 state_dict[fqn_with_prefix] = state_dict.pop(fqn)
+            local_state_dict[fqn_with_prefix] = value
+
+    if info.broadcast_from_rank0:
+        device = None
+        for key, value in local_state_dict.items():
+            if torch.is_tensor(value) and value.dim() > 0:
+                if device is None:
+                    device = value.device
+                else:
+                    assert device == value.device
+        assert device is not None
+        _broadcast_state_dict(state_dict, local_state_dict, device=device)
```

Comment:
@fegin for meta init model, we will see error because of `device=meta`. Is there a way to resolve it? eg passing device from `StateDictOptions`?